target/linux/ramips/patches-5.15/120-08-mips-ralink-get-cpu-rate-from-new-driver-code.patch

```diff
@@ -0,0 +1,101 @@
+From ebcffe9317b3e695357fa9840ce859e030ca877a Mon Sep 17 00:00:00 2001
+From: Sergio Paracuellos <sergio.paracuellos@gmail.com>
+Date: Thu, 16 Mar 2023 10:05:53 +0100
+Subject: [PATCH] mips: ralink: get cpu rate from new driver code
+
+At very early stage on boot, there is a need to set 'mips_hpt_frequency'.
+This timer frequency is a half of the CPU frequency. To get clocks properly
+set we need to call to 'of_clk_init()' and properly get cpu clock frequency
+afterwards. Depending on the SoC, CPU clock index and compatible differs, so
+use them to get the proper clock frm the clock provider. Hence, adapt code
+to be aligned with new clock driver.
+
+Signed-off-by: Sergio Paracuellos <sergio.paracuellos@gmail.com>
+---
+ arch/mips/ralink/clk.c | 61 +++++++++++++++++++++++++++++++++++-------
+ 1 file changed, 52 insertions(+), 9 deletions(-)
+
+--- a/arch/mips/ralink/clk.c
++++ b/arch/mips/ralink/clk.c
+@@ -11,29 +11,72 @@
+ #include <linux/clkdev.h>
+ #include <linux/clk.h>
+ #include <linux/clk-provider.h>
++#include <asm/mach-ralink/ralink_regs.h>
+ 
+ #include <asm/time.h>
+ 
+ #include "common.h"
+ 
+-void ralink_clk_add(const char *dev, unsigned long rate)
++static const char *clk_cpu(int *idx)
+ {
+-	struct clk *clk = clk_register_fixed_rate(NULL, dev, NULL, 0, rate);
+-
+-	if (!clk)
+-		panic("failed to add clock");
+-
+-	clkdev_create(clk, NULL, "%s", dev);
++	switch (ralink_soc) {
++	case RT2880_SOC:
++		*idx = 0;
++		return "ralink,rt2880-sysc";
++	case RT3883_SOC:
++		*idx = 0;
++		return "ralink,rt3883-sysc";
++	case RT305X_SOC_RT3050:
++		*idx = 0;
++		return "ralink,rt3050-sysc";
++	case RT305X_SOC_RT3052:
++		*idx = 0;
++		return "ralink,rt3052-sysc";
++	case RT305X_SOC_RT3350:
++		*idx = 1;
++		return "ralink,rt3350-sysc";
++	case RT305X_SOC_RT3352:
++		*idx = 1;
++		return "ralink,rt3352-sysc";
++	case RT305X_SOC_RT5350:
++		*idx = 1;
++		return "ralink,rt5350-sysc";
++	case MT762X_SOC_MT7620A:
++		*idx = 2;
++		return "ralink,mt7620-sysc";
++	case MT762X_SOC_MT7620N:
++		*idx = 2;
++		return "ralink,mt7620-sysc";
++	case MT762X_SOC_MT7628AN:
++		*idx = 1;
++		return "ralink,mt7628-sysc";
++	case MT762X_SOC_MT7688:
++		*idx = 1;
++		return "ralink,mt7688-sysc";
++	default:
++		*idx = -1;
++		return "invalid";
++	}
+ }
+ 
+ void __init plat_time_init(void)
+ {
++	struct of_phandle_args clkspec;
++	const char *compatible;
+ 	struct clk *clk;
++	int cpu_clk_idx;
+ 
+ 	ralink_of_remap();
+ 
+-	ralink_clk_init();
+-	clk = clk_get_sys("cpu", NULL);
++	compatible = clk_cpu(&cpu_clk_idx);
```

Comment:
@paraka I see that the ath79 target query CPU frequency by directly reading the clock of the `cpu` node[1]. Can we use this method on the ralink target? It seems that we just need to add the clock to the `cpu` node in the device tree.

[1]https://github.com/torvalds/linux/blob/e660abd551f1172e428b4e4003de887176a8a1fd/arch/mips/ath79/setup.c#L244-L261

Reply:
@DragonBluep: I saw that also, yes. Thanks for pointing it out. The problem now is that we have already on-tree cpu's yaml bindings for mips: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/mips/cpus.yaml where we cannot force now to add a clock node there. We have to make an exception depending on SoC compatible in the same way I did for broadcom SoCs for mips-hpt-frequency property: https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/mips/brcm/soc.yaml. Since it was tons of not really related amount of work for this series I end up with this solution and we can back to it later when we try to mainline real dtsi files from openWRT. I have started to document some missing compatibles to make easier this task in the future. See: https://lore.kernel.org/linux-devicetree/20230620100231.1412582-1-sergio.paracuellos@gmail.com/T/#u
I was told to remove clock-names property from those places where it is not really necessary so maybe you can also drop them from this PR.